core/src/commonMain/kotlin/com/xebia/functional/xef/prompt/expressions/Expression.kt

```diff
@@ -0,0 +1,98 @@
+package com.xebia.functional.xef.prompt.expressions
+
+import com.xebia.functional.xef.auto.CoreAIScope
+import com.xebia.functional.xef.auto.PromptConfiguration
+import com.xebia.functional.xef.llm.ChatWithFunctions
+import com.xebia.functional.xef.llm.models.chat.Message
+import com.xebia.functional.xef.llm.models.chat.Role
+import com.xebia.functional.xef.prompt.experts.ExpertSystem
+import io.github.oshai.kotlinlogging.KLogger
+import io.github.oshai.kotlinlogging.KotlinLogging
+
+class Expression(
+  private val scope: CoreAIScope,
+  private val model: ChatWithFunctions,
+  val block: suspend Expression.() -> Unit
+) {
+
+  private val logger: KLogger = KotlinLogging.logger {}
+
+  private val messages: MutableList<Message> = mutableListOf()
+
+  private val generationKeys: MutableList<String> = mutableListOf()
+
+  suspend fun system(message: suspend () -> String) {
+    messages.add(Message(role = Role.SYSTEM, content = message(), name = Role.SYSTEM.name))
+  }
+
+  suspend fun user(message: suspend () -> String) {
+    messages.add(Message(role = Role.USER, content = message(), name = Role.USER.name))
+  }
+
+  suspend fun assistant(message: suspend () -> String) {
+    messages.add(Message(role = Role.ASSISTANT, content = message(), name = Role.ASSISTANT.name))
+  }
+
+  fun prompt(key: String): String {
+    generationKeys.add(key)
+    return "{{$key}}"
+  }
+
+  suspend fun run(
+    promptConfiguration: PromptConfiguration = PromptConfiguration.DEFAULTS
+  ): ExpressionResult {
+    block()
+    val instructionMessage =
+      Message(
+        role = Role.USER,
+        content =
+          ExpertSystem(
+              system = "You are an expert in replacing variables in templates",
+              query =
+                """
+          |I want to replace the following variables in the following template:
+          |<template>
+          |${messages.joinToString("\n") { it.content }}
+          |</template>
+          |The variables are:
+          |${generationKeys.joinToString("\n") { it }}
+        """
+                  .trimMargin(),
+              instructions =
+                listOf(
+                  "Create a `ReplaceKeys` object `replacements` property of type `Map<String, String>` where the keys are the variable names and the values are the values to replace them with.",
```

Comment:
Suggestion:

Instead of using Map String String and single value replacement, why not use an expert in a popular templating language, such as Handlebars? This would allow for more complex json object substitutions, loops for lists, etc.

As long as the templating language is popular enough, I think the LLM will be able to come up with a valid template for it.

Reply:
I tried the handlebars syntax support but the LLM is more precise when I ask it to place the structure back on the map than when I ask it to replace all the placeholders in the text, but with some prompt tweaking, it would support handlebars syntax. I still like that we use functions like `prompt` to place the placeholders where generation would happen instead of just providing an entire template and asking it to infer the keys. This allows us to save the keys and ask directly to translate those keys. We could do that with handlebars and some parsing too. Still, I also think you can support the handlebars' control structures and syntax simultaneously if we provide a way to know the keys and regions before we ask the LLM. If you'd like to play with it in the same branch we can try to evolve it to have more powerful templating.

Reply:
I'm thinking something similar to this: 

https://github.com/microsoft/guidance

Granted, many of the features are similar to xef, but specifically In speaking about the templating language triggering the appropriate agents and structured completions here.

It's something I think we can expand in follo-on features,  certainly.